### PR TITLE
[MM-387]: Added support to subscribe to release and deployment

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**
### Description
Added support to subscribe to release and deployment.

### Screenshots
release7AugGitlab
image (2)

### What to Test
- create a subscription with release and deployment features and test the expected notifications are received.

### CODE REVIEW

File: server/webhook/deployment.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	handlers, err := w.handleChannelDeployment(ctx, event)
+	if err != nil {
+		return nil, err
+	}
+	return cleanWebhookHandlers(handlers), nil
+}
+
+func (w *webhook) handleChannelDeployment(ctx context.Context, event *gitlab.DeploymentEvent) ([]*HandleWebhook, error) {
+	senderGitlabUsername := event.User.Username
+	repo := event.Project
```
Comment (raghavaggarwal2308) [inline server/webhook/deployment.go:20]:
```suggestion
	project := event.Project
```

File: server/webhook/deployment.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+		message += fmt.Sprintf(":rocket: **Status**: %s\n", event.Status)
+	case statusCreated:
+		message += fmt.Sprintf(":clock1: **Status**: %s\n", event.Status)
+	case statusSuccess:
+		message += fmt.Sprintf(":large_green_circle: **Status**: %s\n", event.Status)
+	case statusFailed:
+		message += fmt.Sprintf(":red_circle: **Status**: %s\n", event.Status)
+	default:
+		return res, nil
+	}
```
Comment (raghavaggarwal2308) [inline server/webhook/deployment.go:35]:
```suggestion
	}
	
```

File: server/webhook/deployment.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	namespaceMetadata, err := normalizeNamespacedProjectByHomepage(event.Project.Homepage)
+	if err != nil {
+		return nil, err
+	}
```
Comment (raghavaggarwal2308) [inline server/webhook/deployment.go:39]:
```suggestion
	}
	
```

File: server/webhook/deployment.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	fullNamespacePath := fmt.Sprintf("%s/%s", namespaceMetadata.Namespace, namespaceMetadata.Project)
+	message += fmt.Sprintf("**Repository**: [%s](%s)\n", fullNamespacePath, event.Project.GitHTTPURL)
+	message += fmt.Sprintf("**Triggered By**: %s\n", senderGitlabUsername)
+	message += fmt.Sprintf("**Visit job [here](%s)** \n", event.DeployableURL)
+	toChannels := make([]string, 0)
+	subs := w.gitlabRetreiver.GetSubscribedChannelsForProject(
+		ctx, namespaceMetadata.Namespace, namespaceMetadata.Project,
+		repo.VisibilityLevel == PublicVisibilityLevel,
+	)
```
Comment (raghavaggarwal2308) [inline server/webhook/deployment.go:48]:
```suggestion
	subs := w.gitlabRetreiver.GetSubscribedChannelsForProject(
		ctx, namespaceMetadata.Namespace,
		namespaceMetadata.Project,
		repo.VisibilityLevel == PublicVisibilityLevel,
	)
```

File: server/webhook/deployment.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	for _, sub := range subs {
+		if !sub.Deployments() {
+			continue
+		}
+
+		toChannels = append(toChannels, sub.ChannelID)
+	}
```
Comment (raghavaggarwal2308) [inline server/webhook/deployment.go:55]:
```suggestion
	}
	
```

File: server/webhook/deployment.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	return cleanWebhookHandlers(handlers), nil
+}
+
+func (w *webhook) handleChannelDeployment(ctx context.Context, event *gitlab.DeploymentEvent) ([]*HandleWebhook, error) {
+	senderGitlabUsername := event.User.Username
+	repo := event.Project
+	res := []*HandleWebhook{}
+	message := fmt.Sprintf("### Deployment Stage: **%s**\n", event.Status)
+
+	switch event.Status {
```
Comment (raghavaggarwal2308) [inline server/webhook/deployment.go:24]: Add a screenshot for each type of event we are receiving in the description

File: server/webhook/release.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return cleanWebhookHandlers(handlers), nil
+}
+
+func (w *webhook) handleChannelRelease(ctx context.Context, event *gitlab.ReleaseEvent) ([]*HandleWebhook, error) {
+	repo := event.Project
+	res := []*HandleWebhook{}
+	message := fmt.Sprintf("### Release: **%s**\n", event.Action)
```
Comment (raghavaggarwal2308) [inline server/webhook/release.go:21]: Dont you think it will be better to use a slack attachment here instead of markdown?

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: server/webhook/release.go [hunk 8]
```diff
@@ -16,7 +16,7 @@ func (w *webhook) HandleRelease(ctx context.Context, event *gitlab.ReleaseEvent)
 }
 
 func (w *webhook) handleChannelRelease(ctx context.Context, event *gitlab.ReleaseEvent) ([]*HandleWebhook, error) {
-	repo := event.Project
+	release := event.Project
```
Comment (raghavaggarwal2308) [inline server/webhook/release.go:19]: Shouldnt this be project?

**Comment by raghavaggarwal2308** [comment]
Also, fix the failing lint

**Comment by Kshitij-Katiyar** [comment]
Created a PR with all the changes in the Mattermost organization's GitLab plugin repo, so closing this one

**Pull request closed without merging**